File: core/src/main/java/com/xmartlabs/bigbang/core/controller/SessionController.java
package com.xmartlabs.bigbang.core.controller;

import android.content.SharedPreferences;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

import com.annimon.stream.Optional;
import com.google.gson.Gson;
import com.xmartlabs.bigbang.core.model.SessionType;

import javax.inject.Inject;

import io.reactivex.Single;

/**
 * Controller that manages the Session of the Application.
 *
 * The Session will be stored in the {@link SharedPreferences} as a String.
 * Thus, the {@link SessionType} object is serialized using {@link Gson}.
 * The session is retrieved once from disk and then kept in memory for faster access.
 */
public abstract class SessionController extends Controller {
  private static final String PREFERENCES_KEY_SESSION = "session";

  @Inject
  Gson gson;
  @Inject
  SharedPreferences sharedPreferences;

  private Optional<SessionType> session = Optional.empty();

  /**
   * Retrieves the current stored {@link SessionType}, if it exists.
   *
   * Only upon first request the {@link SessionType} object will be queried from {@link SharedPreferences}.
   * Then, it will be stored in memory for faster access.
   *
   * @return the current {@link SessionType}, or {@code null} if none exists
   */
  @CheckResult
  @NonNull
  public Optional<SessionType> getAbstractSession() {
    return Optional.of(session).orElseGet(() -> {
      String sessionJsonString = sharedPreferences.getString(PREFERENCES_KEY_SESSION, null);
      return deserializeSession(sessionJsonString);
    });
  }

  protected abstract Optional<SessionType> deserializeSession(String json);

  /**
   * Stores the {@code session} into the {@link SharedPreferences}.
   *
   * @param session the {@link SessionType} object to be stored
   * @return {@code Single<Session>} object. Upon subscription, it will only fail if the session could not be stored
   */
  @CheckResult
  @NonNull
  public <S extends SessionType> Single<S> setSession(@NonNull S session) {
    return Single.fromCallable(() -> {
      String sessionJsonString = gson.toJson(session);
      boolean committed = sharedPreferences
          .edit()
          .putString(PREFERENCES_KEY_SESSION, sessionJsonString)
          .commit();

      if (committed) {
        this.session = Optional.of(session);
        return session;
      }
      throw new RuntimeException("The session change could not be committed");
    });
  }

  /** Deletes the session information */
  public void deleteSession() {
    session = Optional.empty();
    sharedPreferences
        .edit()
        .remove(PREFERENCES_KEY_SESSION)
        .apply();
  }

  /**
   * Returns whether the {@link SessionType} information is present on the device.
   * @return whether or not the {@link SessionType} information exists
   */
  @CheckResult
  @SuppressWarnings("unused")
  public boolean isSessionAlive() {
    return sharedPreferences
        .contains(PREFERENCES_KEY_SESSION);
  }
}

File: core/src/main/java/com/xmartlabs/bigbang/core/controller/SessionController.kt
package com.xmartlabs.bigbang.core.controller

import android.content.SharedPreferences
import android.support.annotation.CheckResult
import com.xmartlabs.bigbang.core.model.SessionType
import javax.inject.Inject
import kotlin.reflect.KClass

/**
 * Controller that manages the Session of the Application.
 *
 * The Session will be stored via the [SharedPreferencesController].
 */
open class SessionController(private val sessionType: KClass<out SessionType>) : Controller() {
  companion object {
    internal val PREFERENCES_KEY_SESSION = "session"
  }
  
  @Inject
  protected lateinit var sharedPreferencesController: SharedPreferencesController

  /**
   * Retrieves the current stored [SessionType], if it exists.
   *
   * @return the current [SessionType], or `null` if none exists
   */
  open val abstractSession
      get() = sharedPreferencesController.getEntity(getSessionKey(), sessionType)
  /**
   * Returns whether the [SessionType] information is present on the device.
   *
   * @return whether or not the [SessionType] information exists
   */
  open val isSessionAlive
    @CheckResult
    get() = sharedPreferencesController.hasEntity(getSessionKey())

  /**
   * Stores the `session` into the [SharedPreferences].
   *
   * @param session the `S` object to be stored
   * *
   * @param <S> the [SessionType] object to be stored
   * */
  open fun <S : SessionType> saveSession(session: S) =
      sharedPreferencesController.saveEntity(getSessionKey(), session)
  
  /** Deletes the session information  */
  open fun deleteSession() = sharedPreferencesController.deleteEntity(getSessionKey())
  
  open protected fun getSessionKey() = PREFERENCES_KEY_SESSION
}
